### DISCUSSION TITLE
How to rate limit server actions via middleware?

### DISCUSSION THREAD

**Discussion by sarimrmalik** [vercel/next.js#62557] [Help]
### Summary

Wondering how to rate limit server actions using Middleware in Nextjs 14?

You can match API routes using the `'/api/:function*'` [path](https://nextjs.org/docs/app/building-your-application/routing/middleware#producing-a-response), wondering if there is a specific convention for server actions?

I understand that I could add rate limiting to the server action itself, however, I want to prevent the server action from being invoked in the first place. Would love any help on this.

Link to [middleware](https://github.com/RubricLab/maige/blob/main/middleware.ts) under discussion.

**Comment by boncz92** [comment]
While not facing the same issue, I am facing an issue with returning a response from the action in middleware, when you return a response the response is not accessible from the client side as it is expecting a rsc payload and not a json response. Wondering if you had any luck finding a solution to this.

**Reply by xcharleslin** [reply]
Looks like they've recently fixed this in https://github.com/vercel/next.js/pull/67059

**Comment by versecafe** [comment]
@sarimrmalik you can check for the Header called `Next-Action` and then apply your rate limiter over all requests with that header, actions will get a unique ID per deployment which you can then use to make your rate limiter specific to limit x number of y action over time z

<img width="343" alt="image" src="https://github.com/vercel/next.js/assets/147033096/8f6f0039-55fb-4277-84de-3495b45be612">

**Reply by ThomasBurgess2000** [reply]
How do you define the map of actions to rate limit prior to deployment if the IDs are unique per deployment?

**Reply by versecafe** [reply]
@ThomasBurgess2000 
Actions do go by route so you can rate limit POST requests to the route where action is called, while this does apply the rate limiter over the entire route. While not as fine grained as limiting action A and action B on the exact same page it does cover most use cases

**Reply by simovicaleksa** [reply]
Can you explain what do you mean by: "Actions do go by route so you can rate limit POST requests to the route where action is called".

I am currently doing something like this:

```ts
export async function middleware(request: NextRequest) {
  // check if the request is a server action and apply the rate limiting then
  if (typeof request.headers.get("Next-Action") === "string")
    // ... rate limiting logic
  }

export const config = {
  matcher: "/"
}
```

How to distinguish between the requests besides the id that is inconsistent between deployments?

Also, is there any way to send a user on the front-end some kind of feedback when the middleware blocks the server action due to rate limiting?

**Reply by boncz92** [reply]
So currently you can’t do a response via middleware but that’s fixed in ndxtjs15 but you can in middleware attach a header called mw-ratelimit and said it to true and key in off that in the actions, that’s currently what I do. 

The other key to this solution that you mentioned is to cache the route id that’s generated at build time and who is running it and so on. You essentially have to build your own rate limiting using a cache. I don’t know of a better solution right now I don’t rate limit I just verify they are authenticated if it’s a server action and no from the login page and let it go if they are authenticated.

**Reply by versecafe** [reply]
@Aleksa1312 the actions generate a POST request endpoint matching the route of the page where they are used so an action on `/settings/account` would have a POST to `/settings/account` with a `Next-Action` header, between those two you can do a decent bit of limiting

**Reply by tn02103** [reply]
Hey @versecafe  you should never use the route for a ServerAction to do anything like Authentication or RequestLimiting. Yes the ServerAction will have the route from the page where it is used, but the problem is that the route is not validated by NextJs. (At least not in the version 14, with verstion 15 I'm not sure) So you can actually just intercept and manipulate the request and just change the route to something like `/login` and NextJs will not care.
